### ISSUE TITLE
Pipelined Rendering: Deferred Extraction Commands

### ISSUE THREAD

**Issue by james7132** [bevyengine/bevy#5082]
## What problem does this solve or what need does it fill?
Running commands can get expensive. They require significant amounts of time relative to just normal system query iteration and require exclusive access to the World they're being run on. Extract in particular currently has an exclusive hold on *both* main and render worlds and takes significant amounts of time in large or complex scenes.

## What solution would you like?
When pipelined rendering is enabled, instead of holding an exclusive lock on both worlds. Run read-only over the main world, and send the commands to the render world, deferring command application to run in parallel with the next game tick's game update.

![image](https://user-images.githubusercontent.com/3137680/175370127-31528a6a-4c1c-433c-a17f-ebe7610a7c73.png)

A benefit of this approach is that there is no explicit synchronization point here other than the start of the frame. The only downside is that we need a secondary storage to drain all CommandQueues into a single one to send to the Render World. This last part can be exceptionally fast as it's essentially a single `memcpy` per system.

## What alternative(s) have you considered?
Explicit synchronization point between main and render worlds.

**Comment by superdump** [comment]
How would backpressure be applied to simulation? What if simulation can run at 1000fps but we can only render at 30fps?

In general I feel like we should be trying to avoid having to apply lots of commands that take significant time due to causing the highest frequency entities (meshes, sprites) to be repeatedly moved between ECS tables and instead try to do as much in-place and in-parallel as we can. I think if we do that, then maybe the first frame isn’t super fast or when a lot of entities are spawned or perhaps come into view, but the baseline should be fast. And even when inserting a bunch of new entities, if we avoid the moving between tables, we should still be able to do extraction pretty fast. I think preallocation, fast allocation, in-place writing of component data (not the current heap allocate and push components to vecs for each mesh/sprite/any extraction system to do a batch insertion except there are multiple of these) is the real solution to the problem of slow extraction. Don’t you? And @cart?

**Comment by james7132** [comment]
> How would backpressure be applied to simulation? What if simulation can run at 1000fps but we can only render at 30fps?

Both are synchronized at the start of a frame, so if the simulation can run at 1ms but the render takes 32ms, the main thread will wait until the render is available. As with normal pipelined rendering ,the game tick time is bounded by the slower of the two. This also gives more time to other intensive operations. If we have rollback networking, it may not be unusual to see the main world running 2-15 times faster than rendering.

> In general I feel like we should be trying to avoid having to apply lots of commands that take significant time due to causing the highest frequency entities (meshes, sprites) to be repeatedly moved between ECS tables and instead try to do as much in-place and in-parallel as we can. 

I am tempted to agree there. Though in the case we *do* have a large update swath in a given frame, the explicit synchronization slows both the simulation and rendering instead of just one or the other. With that said, there is ample opportunity for optimizing commands. This can be seen with #5074, where deduplicating/collapsing archetype moves across commands also cuts down on the large components being moved from table to table.

> I think preallocation, fast allocation, in-place writing of component data (not the current heap allocate and push components to vecs for each mesh/sprite/any extraction system to do a batch insertion except there are multiple of these)

In a way, this *is* doing the same thing as preallocating and writing to a change-detection-less RenderWorld. It's equivalent to writing a singular huge column/table that doesn't have the fragmentation issues of normal ECS storage, and is a copy/write once, leave unchanged operation. The original Destiny talk on this kind of architecture explicitly stated that the primary goal was to extract a small, cache-coherent set of buffers that the renderer could easily point into, which is in-line with this kind of approach: we write out a small pointer as a component (`DynamicUniformIndex<T>`) that points into small cache-coherent buffers (`ComponentUniforms<T>`). If we need that information, we can always index back into these buffers if need be.

As it currently stands, we really only use the information stored in `MeshUniform` when queuing material meshes, and only for the transform's pose for distance calculation. The cost of copying a 128+ byte struct multiple times, both as commands for insertion and during table moves in ECS storage is difficult to justify, particularly when it *must* be accompanied by a handle *and* a dynamic uniform index to be rendered properly.

**Comment by maniwani** [comment]
@james7132, is this issue still unresolved? Seems to have been taken care of by #4402.

`Extract` only takes shared borrows on the main world, and we don't even have to copy command queues since the render world is running it. And if it wasn't already, `Extract` should be the only sync point between the worlds once they're on separate threads.
